### PR TITLE
[ADD] Minimal prototype for KFAC

### PR DESCRIPTION
**Pull request by f-dangel**
Progress on #42. 

This adds a first prototype for a linear operator of KFAC. It has very limited support (e.g. only `Linear` layers, no weight sharing), but I think this is a good milestone to start off with (and to keep the `git diff`s manageable. The tests compare the per-layer GGN with KFAC for deep linear models with `MSELoss` for a large number of MC samples from the predictive distribution.

### CODE REVIEW

File: test/conftest.py [hunk 1]
```diff
@@ -56,3 +58,17 @@ def non_deterministic_case(
 @fixture(params=ADJOINT_CASES)
 def adjoint(request) -> bool:
     return request.param
+
+
+@fixture(params=KFAC_EXPAND_EXACT_CASES)
+def kfac_expand_exact_case(
```
Comment (runame) [inline test/conftest.py:64]: I don't fully understand why this function is necessary, could you briefly explain?

**Review by runame — APPROVED** [review]
LGTM! I think this is a good starting point and we can start adding the missing cases/functionality. Maybe we can create issues of the missing cases and distribute them among the two of us?

File: test/conftest.py (hunk 1 above)
Comment (f-dangel) [reply test/conftest.py:64]: Functions in `conftest.py` are fixtures, i.e. they can be used by test functions. This happens below where `kfac_expand_exact_case` is used as argument of a test function. The test will then iterate through all cases yielded by `kfac_expand_exact_case`. This is similar to `mark.parametrize`. One advantage is that we can also do things like fixing the random seed in the fixture, so the setup code does not leak into the test function.

Comment (f-dangel) [reply test/conftest.py:64]: Disclaimer: I experimented with `fixture`s in this project. Not sure if they are the best way though... If you believe this is too complicated, I'm okay adding new tests in a `fixture`-free way.

File: test/kfac_cases.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+from functools import partial
+from test.utils import get_available_devices, regression_targets
+
+from torch import rand
+from torch.nn import Linear, MSELoss, Sequential
+
+# Add test cases here, devices and loss function with different reductions will be
+# added automatically below
+KFAC_EXPAND_EXACT_CASES_NO_DEVICE_NO_LOSS_FUNC = [
```
[the remark this answers is not in the document]
Comment (f-dangel) [reply test/kfac_cases.py:11]: Agree that it does not make sense at the moment. I decided to already name it like this thinking that we will soon support convolution layers (i.e. weight sharing), and the related tests (using the KFC approximation of Grosse 2016) would go here.

File: curvlinops/kfac.py [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+
+        Warning:
+            This is an early proto-type with many limitations:
+
+            - Parameters must be in the same order as the model's parameters.
+            - Only linear layers with bias are supported.
+            - Weights and biases are treated separately.
+            - No weight sharing is supported.
+            - Only the Monte-Carlo sampled version is supported.
+            - Only the ``'expand'`` setting is supported.
```
[the remark this answers is not in the document]
Comment (f-dangel) [reply curvlinops/kfac.py:101]: Assume we support KFAC from Martens 2015 and KFC from Grosse 2016. Then this is the `'expand'` approximation, right? You're correct that for linear layers without weight sharing, `expand=reduce`.

File: curvlinops/kfac.py [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+
+            batch_size = g.shape[0]
+            correction = {
+                "sum": 1.0 / self._mc_samples,
+                "mean": batch_size**2 / (self._N_data * self._mc_samples),
+            }[self._loss_func.reduction]
+            covariance = einsum("bi,bj->ij", g, g).mul_(correction)
+        else:
+            # TODO Support convolutions
+            raise NotImplementedError(f"Layer of type {type(module)} is unsupported.")
```
[the remark this answers is not in the document]
Comment (f-dangel) [reply curvlinops/kfac.py:324]: Done.

File: test/test_kfac.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+from curvlinops.ggn import GGNLinearOperator
+from curvlinops.kfac import KFACLinearOperator
+
+
+def test_kfac(
+    kfac_expand_exact_case: Tuple[
+        Module, MSELoss, List[Parameter], Iterable[Tuple[Tensor, Tensor]]
+    ]
+):
+    model, loss_func, params, data = kfac_expand_exact_case
```
[the remark this answers is not in the document]
Comment (f-dangel) [reply test/test_kfac.py:20]: Right now, `kfac_expand_exact_case` is a `pytest` `fixture` (see `conftest.py`). This means that the function will iterate over all cases yielded by `kfac_expand_exact_case()`, then execute. I don't think the un-packing can be done before. So either we refactor the `fixture` into a `mark.parametrize`, or leave it as is.

File: test/conftest.py (hunk 1 above)
Comment (runame) [reply test/conftest.py:64]: I see, thanks! I'm fine with this design, not sure if the advantages are worth the additional complexity, but I'm happy to try it for this project. We can still simplify later.

File: test/kfac_cases.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (runame) [reply test/kfac_cases.py:11]: Ok, I guess the question is if we should have three sets of test cases: `KFAC`, `KFAC-expand`, and `KFAC-reduce`. I think it probably makes sense to also explicitly test inputs without weight sharing.

File: curvlinops/kfac.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (runame) [reply curvlinops/kfac.py:101]: Yeah, KFC is the expand approximation. And a nitpick: it should also be "approximation" instead of "setting" here.

File: test/test_kfac.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (runame) [reply test/test_kfac.py:20]: Makes sense with the context of the other comment. Let's just stick to using `fixture`s.

File: test/kfac_cases.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (runame) [reply test/kfac_cases.py:11]: To clarify this comment, the current test cases would just be `KFAC`. As soon as we support convolutions, we would only add them to the `KFAC_EXPAND´ (and later `KFAC_reduce`) test cases, together with `Linear` modules with inputs with weight sharing.

[the remark this answers is not in the document]
Comment (f-dangel) [reply test/kfac_cases.py:11]: Let's stick to the more specific naming for now and refactor once we see the need.
